modules/edu.gemini.seqexec.web/edu.gemini.seqexec.web.server/src/main/scala/edu/gemini/seqexec/web/server/security/HttpAuthenticator.scala

```diff
@@ -0,0 +1,63 @@
+package edu.gemini.seqexec.web.server.security
+
+import edu.gemini.seqexec.model.UserDetails
+import edu.gemini.seqexec.web.server.security.AuthenticationService.AuthResult
+
+import org.http4s.{AuthedService, Request, Cookie}
+import org.http4s.dsl._
+import org.http4s.headers
+import org.http4s.server.AuthMiddleware
+
+import scalaz._
+import Scalaz._
+import scalaz.concurrent.Task
+
+import java.time.Instant
+
+/**
+  * Bridge between http4s authentication and the actual internal authenticator
+  */
+class Http4sAuthentication(auth: AuthenticationService) {
+  val cookieService = CookiesService(auth.config.cookieName, auth.config.useSSL, auth.sessionTimeout.toSeconds.toLong)
+
+  def loginCookie(user: UserDetails): Task[Cookie] = cookieService.loginCookie(auth, user)
+
+  val authUser: Kleisli[Task, Request, AuthResult] = Kleisli({ request =>
+    val authResult = for {
+      header <- headers.Cookie.from(request.headers).toRightDisjunction(MissingCookie)
+      cookie <- header.values.list.find(_.name === auth.config.cookieName).toRightDisjunction(MissingCookie)
+      user   <- auth.decodeToken(cookie.content)
+    } yield user
+    Task.delay(authResult)
+  })
+
+  val optAuth = AuthMiddleware(authUser)
+
+  val onFailure: AuthedService[AuthenticationFailure] = Kleisli(req => Forbidden())
+  val reqAuth = AuthMiddleware(authUser, onFailure)
+
+}
+
+trait CookiesService {
+  def name: String
+  def ssl: Boolean
+  def ttl: Long
+
+  def buildCookie(token: String): Task[Cookie] = Task.delay {
+    // if successful set a cookie
+    val exp = Instant.now().plusSeconds(ttl)
+    Cookie(name, token, path = "/".some, expires = exp.some, secure = ssl, httpOnly = true)
+  }
```

Comment:
I think only `exp` needs to be wrapped with a `Task` and `Cookie` could map over `exp`.

In this case computing the time and creating the cookie is so instantaneous that I don't think it doesn't really matter, but if the computation of `Cookie.apply` had to take some time to complete, I think it would.